### ISSUE TITLE
머신러닝/지도학습

### ISSUE THREAD

**Issue by Ludobico** [Ludobico/Python_AL_01#15]
# 머신러닝 알고리즘 분류
* 회귀 분석(지도학습)
    * Linear Regression

* 분류 분석(지도학습)
    * Logistic Regression
    * K-Nearest Neighbor
    * Decision Tree
    * Random Forest
    * Naive Bayes Classifier
    * SVM

* 군집분석(clustering)(비지도 학습)
    * K-Means Clustering
    * Hierarchical Clustering
    * Density Clustering

## 알고리즘 설명
* 선형회귀모델은 2차원 상의 일직선이므로 다음과 같은 다항함수로 표현할 수 있다.
####  y = wx + b

* 모델에 특정한 값을 입력했을 때 얻어지는 결과를 y^ 라고한다. y^는 모델이 만들어낸 예측값이다.
### Linear Regression 종류
* 단순 선형 회귀
    * 단순 선형 회귀는 독립변수가 1개인 선형 회귀를 말한다. y = wx + b

* 다중 선형 회귀
    * feature가 2개 이상인 선형 회귀를 말한다. y = w1x1 + w2x2 + ... +b

### Cost function
* 실제값 y와 예측값 y^의 차이를 오차(error)라고 한다.
* 기계학습은 실제값과 예측값의 오차(error)를 최소화하는 w와 b를 찾아가는 과정이다.
* w, b와 오차의 관계를 함수로 나타낸 것을 비용함수(cost function) 또는 손실함수(Loss function)이라고 부른다.

### 평균제곱오차(MSE, Mean Squared Errors)
* 오차를 전체적으로 줄여야 예측의 정확도가 높아진다.
* MSE = n/1 * sigma(n)(i = 1) (yi = y^i)

### Optimizer
* 비용함수를 최소화하는 w와 b를 구하는 최적화 알고리즘을 옵티마이저라 부른다.
    * 경사하강법(Gradient Descent)
        * 가장 기본적인 옵티마이저이다.
        * 비용함수의 기울기가 작아지는 방향으로 w와 b를 업데이트 한다.
        * 비용함수를 그래프로 그리면 x ** 2 그래프가 된다.
    

## 다중 회귀식의 추정방법
* 동시 입력법
    * 모든 독립변수들을 포함하여 분석하는 방법
    * 이를 통해 특정 독립변수의 영향력을 알 수 있음
    
* 단계적 선택법
    * 다른 변수들이 회귀식에 존재할 때 종속변수에 영향력이 있는 변수들만을 회귀식에 포함시키는 방법
    * 설명력이 높은 즉, 유의 확률 p가 가장 작은 변수의 순으로 회귀식에 포함시킴
    
* 후진 소거법
    * 모든 독립변수를 모두 포함시킨 상태에서 기여도가 적은 변수부터 하나씩 제거하는 방법
    
* 전진 선택법
    * 독립변수가 하나도 포함되지 않은 모델에서 시작해서 f값에 가장 큰 기여를 하는 변수를 순서대로 더하는 방법

## 다중공선성과 VIF
* 다중공선성이란
    * 회귀분석에서 사용된 모형의 일부 변수가 다른 변수와 상관정도가 높아 데이터 분석시 부정적인 영향을 미치는 현상
    * 다중 공선성이 발생되면, 회귀 모형의 적합성이 떨어지고, 다른 중요한 독립변수가 모형에서 제거 될 가능성이 높다.
    * 결정계수의 값이 과대하게 나타날 수 있거나 설명력은 좋은데 예측력이 떨어질 수 있게 된다.
    * 공선성을 낮추는 방법은 상관관계가 높은 독립변수를 제거하거나 변수 선택방법을 이용해서 분석해야 한다.
    
* 다중공선성 의심 상황
    * data 수에 비해 과다한 독립변수를 사용했을 때
    * 독립변수들의 상관계수가 크게 나타날 때
    * 한 독립변수를 회귀모형에 추가하거나 제거하는 것이 회귀계수의 크기나 부호에 큰 변화를 줄 때

* 다중공선성을 파악하는 방법
    * 산점도 그래프
    * VIF

* VIF(Variance Inflation Factors, 분산 팽창요인)
    * 다중 회귀 모델에서 독립 변수간 상관 관계가 있는지 측정하는 척도
    * VIF가 10이 넘으면 다중공선성이 있다고 판단하며 5가 넘으면 주의할 필요가 있는 것으로 판단한다.
    * 절대적인 기준은 없으므로 적절한 수준에서 판단해야 한다.

## 정규화 선형회귀(Ridge, Lasso, Elastic-net)
![image-2.png](attachment:image-2.png)
* 정규화 선형회귀는 선형회귀 계수에 대한 제약 조건을 추가함으로써 모델이 과적합하는 현상을 막는 방법이다.
* 과적합이 발생하면 모델의 가중치도 과하게 증가하는 경향이 나타나는데, 이에 착안하여 가중치의 크기를 제한함으로써 과적합을 막을 수 있다.
* 과적합은 학습데이터에 대해서는 완벽하게 예측하지만, 새로운 데이터가 들어왔을때는 예측이 안정적이지 않는것이다.
* 단 정규화의 정도를 강하게 할 경우에는 학습이 잘 되지 않는 과소적합이 발생할 수 있다.
* 일반적으로 Ridge regression, Lasso regression, Elastic-net regression이 사용된다.

#### Bias와 Variance
* 예측값들이 실제 타겟값과 전반적으로 멀리 떨어져 있으면 결과의 편향(bias)가 높다고 말한다.
* 예측값들이 전반적으로 서로 멀리 흩어져 있으면 결과의 분산(variance)가 높다고 말한다.
![image.png](attachment:image.png)

#### trade-off
* 편향과 분산은 한쪽이 증가하면 다른 한쪽이 감소하고, 한쪽이 감소하면 다른 한쪽이 증가하는 trade-off 관계이다.
* 정규화는 분산을 낮춰 과적합을 막는 방법이지만 trade-off 관계로 인해서 분산을 낮추는 대신 편향이 높아질 수 있다.

## 1. 릿지회귀(ridge Regression)
* (1) 정의
    * 추정계수의 제곱합을 최소로 하는것.
* (2) 특징
     - 제곱합을 최소화 하기 때문에 계수의 크가 줄어듬 => 과적합이 방지됨

     - 다중공선성이 분산되기 때문에 효과적

     - 딥러닝 loss function에서 L2 panelty로 사용됨
    
## 2. 라쏘회귀(lasso Regression)
* (1) 정의
    * 추정계수의 절댓값 합을 최소로 하는것 (0에 가까워짐)
* (2) 특징
    - 몇몇 기여도가 낮은 계수들이 0으로 수렴하려고 함

    - 반면 기여도가 높은 특정변수의 계수는 커지는 단점이 있음

    - 딥러닝 loss function에서 L1 panelty로 사용됨
    
## 3. 엘라스틱 넷(elastic net)
* (1) 정의
    * 추정계수의 절댓값 합과 제곱합을 동시에 최소로 하는것
* (2) 특징
    - 큰 데이터셋에서 릿지와 라쏘의 효과를 모두 반영하기 좋음 (적은 데이터셋은 효과적음, 릿지 라쏘와 반대)
